go/merkletree2/proof_verifier.go

```diff
@@ -0,0 +1,99 @@
+package merkletree2
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+)
+
+type MerkleProofVerifier struct {
+	TreeConfig
+}
+
+func NewMerkleProofVerifier(t TreeConfig) MerkleProofVerifier {
+	return MerkleProofVerifier{TreeConfig: t}
+}
+
+func (m *MerkleProofVerifier) VerifyInclusionProof(ctx context.Context, kvp KeyValuePair, proof MerkleInclusionProof, expRootHash Hash) error {
```

Comment:
In general I think this MerkleInclusionProof struct makes this function a lot harder than it could be. For example you could have sibling hashes be a list of left, right pairs of lists, so you don't need to do any index arithmetic.
And OtherPairsInLeaves could just be the entire set of KV pairs in the leaf so you don't need to manually insert it.

I think the space increase of the proof would be minimal and this function might even get a bit faster without all the copying.

Reply:
I agree the algorithm is not the most readable and there is some bit fiddling, but I'd rather have a smaller proof than a simpler to read algorithm. Do not feel strongly though. @maxtaco ?